abi/src/main/java/org/web3j/abi/TypeDecoder.java

```diff
@@ -579,15 +551,26 @@ private static <T extends Type> T decodeDynamicParameterFromStruct(
             final String input,
             final int parameterOffset,
             final int parameterLength,
-            final Class<T> declaredField) {
+            final Class<T> declaredField,
+            final Class<T> parameter) {
         final String dynamicElementData =
                 input.substring(parameterOffset, parameterOffset + parameterLength);
 
         final T value;
         if (DynamicStruct.class.isAssignableFrom(declaredField)) {
+            value = decodeDynamicStruct(dynamicElementData, 0, TypeReference.create(declaredField));
+        } else if (DynamicArray.class.isAssignableFrom(declaredField)) {
             value =
-                    decodeDynamicStruct(
-                            dynamicElementData, 64, TypeReference.create(declaredField));
+                    (T)
+                            decodeDynamicArray(
+                                    dynamicElementData,
+                                    0,
+                                    new TypeReference<DynamicArray>() {
+                                        @Override
+                                        TypeReference getSubTypeReference() {
+                                            return TypeReference.create(parameter);
+                                        }
```

Comment:
externalise this to a method getType in Utils class because I see it is used in 2 places in your changes, probably even more in the overall code

Reply:
Done. Apparently there were only 2 places with this